src/jwks-fetch.js

```diff
@@ -1,54 +1,67 @@
 'use strict'
 
 const fetch = require('node-fetch')
+const lru = require('tiny-lru')
+
 const MISSING_KEY_ERROR = 'No matching key found in the set.'
-const NO_KEYS_ERROR = 'No keys found in the set.'
 
-async function jwksFetch ({ domain, alg, kid, cache = null }) {
-  const cacheKey = `${alg}:${kid}:${domain}`
+function buildJwksFetch (cacheProps = {}) {
+  const max = cacheProps.max || 100
+  const ttl = cacheProps.ttl || 60 * 1000
+  const cache = lru(max, ttl)
 
-  const cached = cache ? cache.get(cacheKey) : null
+  async function getSecret (signatures) {
+    const { domain, alg, kid } = signatures
+    const cacheKey = `${alg}:${kid}:${domain}`
+    const cachedSecret = cache.get(cacheKey)
 
-  if (cached) {
-    return cached
-  } else if (cache && cached === null) {
-    // null is returned when a previous attempt resulted in the key missing in the JWKs - Do not attemp to fetch again
-    throw new Error(MISSING_KEY_ERROR)
-  }
+    if (cachedSecret) {
+      return cachedSecret
+    } else if (cachedSecret === null) {
+      // null is returned when a previous attempt resulted in the key missing in the JWKs - Do not attemp to fetch again
+      throw new Error(MISSING_KEY_ERROR)
+    }
 
-  // Hit the well-known URL in order to get the key
-  const response = await fetch(`${domain}.well-known/jwks.json`, { timeout: 5000 })
-  const body = await response.json()
+    // Hit the well-known URL in order to get the key
+    const response = await fetch(`${domain}.well-known/jwks.json`, { timeout: 5000 })
+    const body = await response.json()
 
-  if (!response.ok) {
-    const error = new Error(response.statusText)
-    error.response = response
-    error.body = body
+    if (!response.ok) {
+      const error = new Error(response.statusText)
+      error.response = response
+      error.body = body
 
-    throw error
-  }
+      throw error
+    }
 
-  if (!body.keys || body.keys.length === 0) {
-    throw new Error(NO_KEYS_ERROR)
-  }
+    if (!body.keys || body.keys.length === 0) {
+      throw new Error('No keys found in the set.')
```

Comment:
extract in a constant as it was before instead of inlining this

Reply:
yes, thanks!